### PR TITLE
Feature/h5store context manager reentry

### PR DESCRIPTION
**Pull request by csadorf**
Enable reentry of the H5Store context manager.

## Description

This essentially enables:
```python
with job.data:
    with job.data:
```
## Motivation and Context

This is especially useful if you use the store in different functions and just want to make sure that the file is open within that specific context without caring about whether the file is already opened in the outer context.

## Types of Changes

- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [x] Breaking change<sup>1</sup>

This pull request is based on
- [ ] *master*, because it is a bug fix or an update to the documentation.
- [x] *develop*, because it introduces a new feature.

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac/blob/master/contributors.txt).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac/blob/misc/contributor-flow/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [ ] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [ ] I have updated the [changelog](https://github.com/glotzerlab/signac/blob/master/changelog.txt).

### CODE REVIEW

File: signac/core/h5store.py [hunk 1]
```diff
@@ -247,7 +247,11 @@ def __del__(self):
         self.close()
 
     def __enter__(self):
-        self.open()
+        try:
+            self.open()
+        except OSError as error:
+            if 'is already open' not in str(error):
```
Comment (bdice) [inline signac/core/h5store.py]: I'm not sure that I like tying this implementation to the specific error text. I would imagine that non-English versions of Python might not raise an error with this string. (Versuch es bitte?) I tried to find what raises this error (h5py? CPython? Something else?) but couldn't locate the source. Is there another way?

Comment (csadorf) [reply signac/core/h5store.py]: Unfortunately the h5py package overall relies on generic RuntimeErrors for almost all error conditions. I don't know of a better way to limit the scope here. I don't thing that localized versions of the h5py package or the HDF5 library exist and therefore do not anticipate this to be a problem.

**Comment by csadorf** [comment]
I just noticed the tests failing on PyPy2 and 3. I'll fix that now.

**Comment by csadorf** [comment]
@bdice I've addressed all comments and fixed all issues.

**Pull request merged**